common/src/Microsoft.Azure.IIoT.Agent.Framework/src/Agent/Default/Worker.cs

```diff
@@ -223,72 +220,66 @@ public class Worker : IWorker, IDisposable {
+                    _logger.Error(ex, "Worker {workerId}, exception during worker processing, wait {delay}...",
                         WorkerId, _jobCheckerInterval);
                     kModuleExceptions.WithLabels(AgentId, ex.Source, ex.GetType().FullName, ex.Message, ex.StackTrace, "Exception during worker processing").Inc();
                     await Task.Delay(_jobCheckerInterval, ct).ConfigureAwait(false);
                 }
             }
-            _logger.Information("Worker stopping...");
+            _logger.Information("Worker {workerId} stopping...", WorkerId);
         }
 
         /// <summary>
         /// Process jobs
         /// </summary>
-        /// <returns></returns>
         private async Task ProcessAsync(JobProcessingInstructionModel jobProcessInstruction,
             CancellationToken ct) {
             try {
                 // Stop worker heartbeat to start the job heartbeat process
                 _heartbeatTimer.Change(Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan); // Stop worker heartbeat
 
-                _logger.Information("Worker: {WorkerId} processing job: {JobId}, mode: {ProcessMode}",
+                _logger.Information("Worker {WorkerId} processing job {JobId}, mode: {ProcessMode}",
                     WorkerId, jobProcessInstruction.Job.Id, jobProcessInstruction.ProcessMode);
 
                 // Execute processor
                 while (true) {
-                    _jobProcess = null;
                     ct.ThrowIfCancellationRequested();
-                    using (_jobProcess = new JobProcess(this, jobProcessInstruction,
-                        _lifetimeScope, _logger)) {
-                        await _jobProcess.WaitAsync(ct).ConfigureAwait(false); // Does not throw
+                    if (_jobProcess == null) {
+                        _jobProcess = new JobProcess(this, jobProcessInstruction, _lifetimeScope, _logger);
                     }
+                    else {
+                        _jobProcess.ProcessNewInstruction(jobProcessInstruction);
+                    }
+                    await _jobProcess.WaitAsync(ct).ConfigureAwait(false); // Does not throw
 
                     // Check if the job is to be continued with new configuration settings
-                    if (_jobProcess.JobContinuation == null) {
+                    if ((_jobProcess?.JobContinuation?.Job?.JobConfiguration == null ||
+                        _jobProcess?.JobContinuation?.ProcessMode == null) &&
+                        _jobProcess.Status != WorkerStatus.Stopped) {
                         _jobProcess.Status = WorkerStatus.Stopped;
                         await SendHeartbeatWithoutResetTimerAsync().ConfigureAwait(false);
+                        _jobProcess.Dispose();
                         _jobProcess = null;
                         break;
                     }
 
                     jobProcessInstruction = _jobProcess.JobContinuation;
```

Comment:
Why do we assign a new value to input `jobProcessInstruction` parameter ? If it is only for logging, I would suggest to use `_jobProcess.JobContinuation` instead.

Reply:
I did not touch that. the PR was intended to fix an issue and I tried to keep the refactoring minimal. we can decide to refactor more in some other PR.

Reply:
that would crash since _jobProcess is disposed at that point in time ...